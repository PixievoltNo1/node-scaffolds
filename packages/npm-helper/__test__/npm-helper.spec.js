const { locateNearestFilepath } = require('@guanghechen/locate-helper')
const fs = require('fs-extra')
const path = require('path')
const {
  collectAllDependencies,
  createDependencyFields,
  detectMonorepo,
  detectPackageAuthor,
} = require('../index')

// Resolve fixture filepath
const resolveFixturePath = p => path.join(__dirname, 'fixtures', p)

describe('util', function () {
  test('detectMonorepo', async function () {
    expect(detectMonorepo(__dirname)).toBeTruthy()

    const lernaFilepath = locateNearestFilepath(__dirname, 'lerna.json')
    const bakFilepath = lernaFilepath + '.bak'

    if (lernaFilepath != null) {
      fs.moveSync(lernaFilepath, bakFilepath)
    }

    try {
      expect(detectMonorepo(resolveFixturePath('monorepo1'))).toBeTruthy()
      expect(detectMonorepo(resolveFixturePath('monorepo2'))).toBeTruthy()
      expect(detectMonorepo(resolveFixturePath('normal-repo'))).toBeFalsy()
    } finally {
      if (lernaFilepath != null) {
        fs.moveSync(bakFilepath, lernaFilepath)
      }
    }
  })

  test('detectPackageAuthor', function () {
    expect(detectPackageAuthor(__dirname)).toEqual('guanghechen')
    expect(detectPackageAuthor(resolveFixturePath('monorepo1'))).toEqual('guanghechen')
    expect(detectPackageAuthor(resolveFixturePath('monorepo2'))).toEqual('waw')
    expect(detectPackageAuthor(resolveFixturePath('normal-repo'))).toBeNull()
  })
})

describe('dependency', function () {
  test('createDependencyFields', function () {
    expect(createDependencyFields()).toEqual([
      'dependencies',
      'optionalDependencies',
      'peerDependencies',
    ])
  })

  test('collectAllDependencies', function () {
    expect(collectAllDependencies(path.join(__dirname, '../package.json'))).toMatchSnapshot(
      'current repo',
    )
    expect(
      collectAllDependencies(
        resolveFixturePath('monorepo2/package.json'),
        undefined,
        ['rollup'],
        () => true,
      ),
    ).toMatchSnapshot('monorepo2')

    const warningDataList = []
    const warnSpy = jest
      .spyOn(console, 'warn')
      .mockImplementation((...args) => void warningDataList.push(...args))
    expect(
      collectAllDependencies(resolveFixturePath('normal-repo/package.json'), ['peerDependencies']),
    ).toMatchSnapshot('normal-repo')
    expect(warningDataList).toMatchSnapshot('normal-repo -- warnings')
    warnSpy.mockRestore()
  })
})
