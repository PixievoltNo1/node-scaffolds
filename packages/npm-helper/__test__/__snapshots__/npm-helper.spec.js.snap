// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependency collectAllDependencies: current repo 1`] = `
Array [
  "@guanghechen/locate-helper",
  "fs-extra",
  "graceful-fs",
  "jsonfile",
  "universalify",
]
`;

exports[`dependency collectAllDependencies: monorepo2 1`] = `
Array [
  "@babel/core",
  "@babel/code-frame",
  "@babel/highlight",
  "@babel/helper-validator-identifier",
  "chalk",
  "ansi-styles",
  "color-convert",
  "color-name",
  "supports-color",
  "has-flag",
  "js-tokens",
  "@babel/generator",
  "@babel/types",
  "to-fast-properties",
  "jsesc",
  "source-map",
  "@babel/helper-compilation-targets",
  "@babel/compat-data",
  "@babel/helper-validator-option",
  "browserslist",
  "caniuse-lite",
  "electron-to-chromium",
  "escalade",
  "node-releases",
  "picocolors",
  "semver",
  "lru-cache",
  "yallist",
  "@babel/helper-module-transforms",
  "@babel/helper-environment-visitor",
  "@babel/helper-module-imports",
  "@babel/helper-simple-access",
  "@babel/helper-split-export-declaration",
  "@babel/template",
  "@babel/parser",
  "@babel/traverse",
  "@babel/helper-function-name",
  "@babel/helper-get-function-arity",
  "@babel/helper-hoist-variables",
  "debug",
  "ms",
  "globals",
  "type-fest",
  "@babel/helpers",
  "convert-source-map",
  "safe-buffer",
  "gensync",
  "json5",
  "minimist",
  "rollup",
  "fsevents",
]
`;

exports[`dependency collectAllDependencies: normal-repo -- warnings 1`] = `
Array [
  "cannot find package.json for '@guanghechen/not-existed-repo'",
]
`;

exports[`dependency collectAllDependencies: normal-repo 1`] = `
Array [
  "@guanghechen/not-existed-repo",
  "rollup",
]
`;
